src/coprocessor_v2/raw_storage_impl.rs

```diff
@@ -0,0 +1,320 @@
+    }
+
+    async fn batch_delete(&self, keys: Vec<Key>) -> StorageResult<()> {
+        let ctx = self.context.clone();
+        let cf = engine_traits::CF_DEFAULT.to_string();
+        let (cb, f) = paired_future_callback();
+
+        let res = self.storage.raw_batch_delete(ctx, cf, keys, cb);
+
+        match res {
+            Err(e) => Err(e),
+            Ok(_) => f.await.map_err(StorageErrorShim::from)?,
+        }
+        .map_err(StorageErrorShim::from)?;
+        Ok(())
+    }
+
+    async fn delete_range(&self, key_range: Range<Key>) -> StorageResult<()> {
+        let ctx = self.context.clone();
+        let cf = engine_traits::CF_DEFAULT.to_string();
+
+        let (cb, f) = paired_future_callback();
+
+        let res = self
+            .storage
+            .raw_delete_range(ctx, cf, key_range.start, key_range.end, cb);
+
+        match res {
+            Err(e) => Err(e),
+            Ok(_) => f.await.map_err(StorageErrorShim::from)?,
+        }
+        .map_err(StorageErrorShim::from)?;
+        Ok(())
+    }
+}
+
+/// Helper struct for converting between [`storage::errors::Error`] and
+/// [`coprocessor_plugin_api::StorageError`].
+struct StorageErrorShim(StorageError);
+
+impl From<StorageErrorShim> for StorageError {
+    fn from(err_shim: StorageErrorShim) -> Self {
+        err_shim.0
+    }
+}
+
+impl From<storage::errors::Error> for StorageErrorShim {
+    fn from(error: storage::errors::Error) -> Self {
+        let inner = match *error.0 {
+            // Key not in region
+            storage::errors::ErrorInner::Engine(EngineError(box EngineErrorInner::Request(
+                ref req_err,
+            ))) if req_err.has_key_not_in_region() => {
+                let key_err = req_err.get_key_not_in_region();
+                StorageError::KeyNotInRegion {
+                    key: key_err.get_key().to_owned(),
+                    region: todo!(), // TODO: how to construct region here? We only have region_id
+                    start_key: key_err.get_start_key().to_owned(),
+                    end_key: key_err.get_end_key().to_owned(),
+                }
+            }
+            // Timeout
+            storage::errors::ErrorInner::Engine(EngineError(box EngineErrorInner::Timeout(
+                duration,
+            ))) => StorageError::Timeout(duration),
+            // Other errors are passed as-is inside their `Result` so we get a `&Result` when using `Any::downcast_ref`.
+            _ => StorageError::Other(Box::new(storage::Result::<()>::Err(error))),
+        };
+        StorageErrorShim(inner)
+    }
+}
+
+impl From<Canceled> for StorageErrorShim {
+    fn from(_c: Canceled) -> Self {
+        StorageErrorShim(StorageError::Canceled)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::storage::{lock_manager::DummyLockManager, TestStorageBuilder};
+    use futures::executor::block_on;
+    use kvproto::kvrpcpb::Context;
+
+    #[test]
+    fn test_storage_api() {
```

Comment:
You could use `#[tokio::test]` to avoid writing `block_on`. Refer to https://docs.rs/tokio/0.2.25/tokio/attr.test.html for more information.

Reply:
Ah nice, didn't know that! Makes the code much cleaner :)